### DISCUSSION TITLE
Concurrency support

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#85] [Discussions]
I understand from the disparate notes that `call/cc` is the one feature of Scheme *not* implemented in IronScheme, and I'm keen to understand if the main reason for that is level of effort being high, or simply the fact that access to full multicore programming via .NET makes it more pain than gain to implement? 

As a not-entirely side thought, one of my favorite features of Clojure is the Communicating Sequential Processes (CSP) support it "borrowed" from Go, and I observe with interest that the maintainer of Guile implemented a library on and for Guile called "fibers" that provides CSP to Guile. See:

- [Github repository](https://github.com/wingo/fibers)

- [Blog post describing motivation](https://wingolog.org/archives/2017/06/27/growing-fibers)

- [Wiki Manual](https://github.com/wingo/fibers/wiki/Manual)

- [Video talk where Wingo describes the implementation](https://www.youtube.com/watch?v=7IcI6sl5oBc)

I would be interested to see IronScheme be fully compliant with R6RS and implement `call/cc` per the spec (and I'm really starting to love Scheme and am sad for all the years I didn't learn it); however, I'd be happy (maybe even happier) to see CSP added to it with a nicer integration with .NET concurrency and parallel programming primitives. I am interested to work on it as well but wonder you (@leppie) think of one or both being core features of IronScheme out of the box perhaps?

**Comment by leppie** [comment]
There is a `call/cc` in IronScheme, but it is not a complete implementation. You can only do escape continuations, basically `call/ec` from Racket.

Native implementations dont have limits on the runtime, so it is easy for them to implement call/cc and/or fibers (aka green threads). With .NET (and probably JVM too), this is just not possible.

The only way to implement call/cc completely in .NET is to use CPS in the compiler. I did such an expirement before, and while it does work, it was horribly slow (10 times or more), and all the .NET interop pretty much goes out of the window.

Given that many Schemes already provide a proper call/cc and no .NET interop, I would suggest using them instead. IronScheme's niche is .NET interop.

**Comment by bcalco** [comment]
@leppie Within that niche, I'm sure there are some interesting concurrency extensions to IronScheme that are possible by wrapping the various concurrency libraries available in .NET. Just need to "think differently" about the problem, given that strict imitation of native Scheme's continuation-passing style and fibers a le Guile's implementation is more pain than gain to implement, per se.

**Comment by Zambito1** [comment]
@bcalco You may find this interesting https://github.com/IronScheme/IronScheme/pull/141

Let me know if you have any feedback on it :)

**Comment by tribals** [comment]
Hi everyone! Let me kindly join discussion. 😊 

I'm not much familiar with CLR as a runtime, and I'm concerned that it states itself as "object-oriented runtime", but indeed, today all those control flow manipulation devices are regocgnized and roughly classified into two groups: "plain old" continuations and "delimited ones".

Here we are speaking about "goold old" continuations, but it turns out that you can imlement those in terms of delimeted continuations - to that end, they are "more powerful". See (1).

Delimited continuations are know by various names: shift/reset, prompt, fcontrol in Scheme, and fexprs in LISP.

Finally, "yield" from imperative languages considered "analoguos" to those delimited continuations.  See (2).

To that end, they are similar to exceptions, too, and for example there is implementation of call/cc for Kawa (as thesis), which shows at least the possibility of that imlementation. See (3).

But considering all the above, wouldn't it be better to implement delimited continuations, in terms of yield (which I think exists in CLR), and then - implement "good old" continuations using delimited?

:thinking:

[1]: [Final Shift for Call/cc](https://www.deinprogramm.de/sperber/papers/shift-reset-direct.pdf)

[2]: [Yield: mainstream delimited continuations](https://cs.indiana.edu/~sabry/papers/yield.pdf)

[3]: [First-Class Continuations on the Java Virtual Machine: An Implementation within the Kawa Scheme Compiler](https://andrebask.github.io/thesis/)

**Reply by tribals** [reply]
Fibers in Guile implemented with prompts, BTW.

**Comment by leppie** [comment]
> First-Class Continuations on the Java Virtual Machine: An Implementation within the Kawa Scheme Compiler

Didn't know about that one. Busy reading. Interesting and familiar.

At one point I did have a CPS build, but it was slow, and like said in above, there are issues with 'existing' code. In the end, I decided against it for cleanliness. There are countless other Scheme that can provide that facility, if needed.

> in terms of yield (which I think exists in CLR)

That is just C# syntactic sugar. Nothing exists natively in the CLR.

> Fibers in Guile implemented with prompts, BTW.

Fibers in the CLR would be awesome and fit in well compared to the CPS approach.

**Reply by tribals** [reply]
Seems like implementing (delimited!) continuations with exceptions is rarely novel, and is demonstrated all the way down to now 😁 Here are some examples:

- Oleg Kiselyov's famous [delimcc for ML](https://okmij.org/ftp/continuations/implementations.html) (with quote):

> The abstract view lets us distill a minimalistic API, scAPI, sufficient for implementing multi-prompt delimited control. We argue that a language system that supports **exception** and **stack-overflow handling** supports scAPI.

You may find this one especially interesting:

- [Portable Implementation of Continuation Operators in Imperative Languages by Exception Handling](https://www.researchgate.net/publication/2352888_Portable_Implementation_of_Continuation_Operators_in_Imperative_Languages_by_Exception_Handling)

This in not by accident in that such an implementation (of delimited continuations) boils down to execution stack manipulation. For example, direct implementation of `shift/reset` in Scheme 48 (from [1] of previous comment) does exactly that.

As I'm not familiar with CLR, I'm not sure - does CLR allows those manipulations? To what extent (eg, CRL vs DLR)? But I've found this interesting blog post which talks exactly on the matter:

- https://mattwarren.org/2019/01/21/Stackwalking-in-the-.NET-Runtime/

What do you think?

I'm also interested in writing some code when all the questions would be resolved, but I've little experience in writing Scheme compilers - I didn't write one yet. 😅 But I can be your "hands" as I think I have a good understanding of the topic.

**Reply by tribals** [reply]
Another interesting reading to the list, precisely about CLR:

- https://news.ycombinator.com/item?id=285751 (a lot of links, better to post whole HN thread)

But, again, personally I propose to implement delimited continuations instead.
